File: java/scrabble-score/src/main/java/Scrabble.java
import java.util.Map;
import java.util.TreeMap;
import java.util.stream.Collectors;

public class Scrabble {

  private String phrase;
  private static final Map<Integer, Integer> scores = buildScoreMap();

  public Scrabble(String phrase) {
    this.phrase = phrase == null ? "" : phrase;
  }

  public int getScore() {
    return this.phrase.chars()
      .map(this::scoreLetter)
      .sum();
  }

  private int scoreLetter(int codePoint) {
    return Scrabble.scores.getOrDefault(Character.toUpperCase(codePoint), 0);
  }

  private static Map<Integer, Integer> buildScoreMap() {
    Map<Integer, Integer> scores = new TreeMap<>();
    addLetters(scores, "AEIOULNRST", 1);
    addLetters(scores, "DG", 2);
    addLetters(scores, "BCMP", 3);
    addLetters(scores, "FHVWY", 4);
    addLetters(scores, "K", 5);
    addLetters(scores, "JX", 8);
    addLetters(scores, "QZ", 10);
    return scores;
  }

  private static void addLetters(Map<Integer, Integer> scores, String letters, int score) {
    letters.chars().forEach(letter -> scores.put(letter, score));
  }

}

File: elixir/scrabble-score/scrabble.exs
defmodule Scrabble do

  @raw_letter_values [
    {"aeioulnrst",  1},
    {"dg", 2},
    {"bcmp", 3},
    {"fhvwy", 4},
    {"k", 5},
    {"jx", 8},
    {"qz", 10},
  ]

  @letter_values for {letters, score} <- @raw_letter_values, letter <- String.graphemes(letters), do: {letter, score}, into: %{}

  @doc """
  Calculate the scrabble score for the word.
  """
  @spec score(String.t) :: non_neg_integer
  def score(word) do
    word
    |> String.downcase
    |> String.graphemes
    |> Enum.map(&(Map.get(@letter_values, &1, 0)))
    |> Enum.sum
  end
end
